src/main/java/com/snowplowanalytics/snowplow/tracker/payload/TrackerEvent.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2020-2020 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.tracker.payload;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import com.snowplowanalytics.snowplow.tracker.Subject;
+import com.snowplowanalytics.snowplow.tracker.Tracker;
+import com.snowplowanalytics.snowplow.tracker.constants.Constants;
+import com.snowplowanalytics.snowplow.tracker.constants.Parameter;
+import com.snowplowanalytics.snowplow.tracker.events.*;
+
+/**
+ * A TrackerEvent which allows the TrackerPayload to be filled later. The payload will be
+ * filled by the Emitter in the Emitter thread, using the getTrackerPayload() method.
+ */
+public class TrackerEvent {
+
+    private final Event event;
+    private final Tracker tracker;
+    private final List<TrackerPayload> payloads;
```

Comment:
Is it possible to remove `payloads` as a property of this class?

I think you use it only to cache the result of calling `getTrackerPayloads()`.  But I think you only every call `getTrackerPayloads()` once, so there is no benefit from caching it.  Maybe I have missed something.

Reply:
Yep! I did this initially as I was returning the TrackerEvents on failure and I didn't want to rebuild the payloads if the failed events were retried but then I changed my mind and now return the Event so we don't really expose TrackerEvent.